Content/Items/Misc/Weapons.Earthduster.cs

```diff
@@ -0,0 +1,560 @@
+                case var type when type == ModContent.ProjectileType<SoilgunPearlsandSoil>(): return DustID.Pearlsand;
+                case var type when type == ModContent.ProjectileType<SoilgunSiltSoil>(): return DustID.Silt;
+                case var type when type == ModContent.ProjectileType<SoilgunSlushSoil>(): return DustID.Slush;
+                case var type when type == ModContent.ProjectileType<SoilgunMudSoil>(): return DustID.Mud;
+                case var type when type == ModContent.ProjectileType<SoilgunVitricSandSoil>(): return ModContent.DustType<VitricSandDust>();
+            }
+            return DustID.Dirt;
+        }
+
+        public void ShootSoils(Vector2 position)
+        {
+            if (Main.myPlayer != Projectile.owner)
+                return;
+
+            Item heldItem = owner.HeldItem;
+
+            if (owner.HeldItem.ModItem is Earthduster gun && gun.ammoItem is null)
+            {
+                Projectile.Kill();
+                return;
+            }
+
+            int damage = Projectile.damage;
+
+            float shootSpeed = heldItem.shootSpeed;
+
+            float knockBack = owner.GetWeaponKnockback(heldItem, heldItem.knockBack);
+
+            Vector2 shootVelocity = Utils.SafeNormalize(Projectile.velocity, Vector2.UnitY) * shootSpeed;
+
+            if (Main.myPlayer == Projectile.owner)
+                Projectile.NewProjectile(Projectile.GetSource_FromThis(), position, (shootVelocity.RotatedByRandom(MathHelper.ToRadians(15))) * Main.rand.NextFloat(0.9f, 1.1f), SoilType, damage, knockBack, owner.whoAmI);
+
+            for (float k = 0; k < 50; k++)
+            {
+                float rads = 6.28f * (k / 50f);
+                float x = (float)Math.Cos(rads) * 50;
+                float y = (float)Math.Sin(rads) * 25;
+
+                Dust.NewDustPerfect(position, GetDustType(), new Vector2(x, y).RotatedBy(Projectile.rotation + MathHelper.PiOver2) * 0.055f + (Vector2.UnitX.RotatedBy(Projectile.rotation) * 5f), 0, default, 0.85f).noGravity = true;
+            }
+
+            Vector2 ejectPos = position + new Vector2(-35, -5 * owner.direction).RotatedBy(Projectile.rotation);
+            for (int i = 0; i < 3; i++)
+            {
+                Dust.NewDustPerfect(ejectPos, GetDustType(), (-Projectile.velocity * Main.rand.NextFloat(1f, 3f) + Vector2.UnitY * -Main.rand.NextFloat(1f, 3f)).RotatedByRandom(0.45f), Main.rand.Next(150), default, 1.25f);
+            }
+            shots++;
+            SoundEngine.PlaySound(SoundID.Item11, Projectile.position);
+
+            CameraSystem.Shake += 1;
+
+            rotTimer += MaxShootDelay;
+
+            if (owner.HeldItem.ModItem is Earthduster earthduster)
+            {
+                int type = earthduster.currentAmmoStruct.projectileID; // this code is still bad
+                bool dontConsumeAmmo = false;
+
+                if (owner.magicQuiver && earthduster.ammoItem.ammo == AmmoID.Arrow && Main.rand.NextBool(5))
```

Comment:
Should we introduce a helper method for calculating ammo reduction? This isnt the first time we've had to manually recreate it it seems.